### ISSUE TITLE
Replication with vector clock based sequences - POC

### ISSUE THREAD

**Issue by adamcfraser** [couchbase/sync_gateway#867]
The distributed index approach expects that last_seq/since values used during replication provides vbucket-level sequence numbering (i.e. a vector clock of the sequence for each vbucket).

Based on analysis to date, it appears that compression alone isn't going to be enough to convert this vector clock into a sequence value that's small enough to be used during replication.  At minimum the sequence value should be less than 1K. 

The proposed alternative is to use a persistent lookup map to convert last_seq/since sequences to more manageable values during replication. 

This POC has the following tasks:
-  [x] Benchmark to validate that compression doesn't meet the requirements
- [x] Implement lookup map algorithm
- [x] Benchmark performance of replication using lookup map 
- [x] Validate replication correctness using lookup map

**Comment by ajres** [comment]
I revisited the compression, for purely random data I can keep the key below 5.5k.

But I can't see that it is going to be possible to improve much on this worse case scenario.

Based on what I remember, would the following approach work?

Each writer writes its vBucket  snapshots  with an incremental monotonically increasing key, creating a snapshot index. This should be cheap as there is no write contention on the index counter.

When a new event is received from a vBucket it is assigned a vector clock consisting of the snapshots index of all the other writers, plus the complete in memory clock of the current writer. (alternative also use current writer snapshot index, plus just the new vbucket id and event sequence id).

This would significantly reduce the the size of the revision clock, for practical deployments i.e. where the number of writers is significantly smaller than 1024 (the number of vBuckets).

The resulting vector clock can still be compressed before sending in the changes feed.

**Comment by tleyden** [comment]
Weren't there non persistent options to reduce overall payload sizes?

Also which compression algorithms are being tested?

**Comment by ajres** [comment]
I tired with a custom bitset compression algorithm, I have not tried standard compression algos.

**Comment by adamcfraser** [comment]
I spent some time a few weeks back doing high-level analysis of the compression options, including review of the options that were out there.  The first task on this ticket is to road-test and document that information.  However, my feeling is we'd really need to get it down to less than 1k for it to be usable for replication.  I'm skeptical we can get to that for random data (that's less than 1 byte per vbucket), but digging into that is part of this ticket.

Wayne and I discussed a hash function/lookup approach that seems cleaner than having to construct a key from multiple subkeys.  Basically it's a hash function that allows collisions, with additional lookup information added to the hash key to resolve the collision and identify which value in the lookup doc you're interested in.
